tools/indexer/pkg/backfill/backfill.go

```diff
@@ -0,0 +1,171 @@
+
+			continue
+		}
+		var vpub []byte
+		// Schedule async validator pubkey fetch
+		if ei.ProposerIdx != nil {
+			vctx, vcancel := context.WithTimeout(ctx, 5*time.Second)
+			v, verr := beacon.FetchValidatorPubkey(vctx, httpc, cfg.BeaconBase, *ei.ProposerIdx)
+			vcancel()
+
+			if verr != nil {
+				return fmt.Errorf("validator pubkey fetch failed slot=%d: %w", ei.Slot, verr)
+			} else if len(v) > 0 {
+				vpub = v
+				if err := db.UpdateValidatorPubkey(ctx, ei.Slot, vpub); err != nil {
+					return fmt.Errorf("validator pubkey update failed slot=%d: %w", ei.Slot, err)
+				}
+			}
+		}
+		select {
+		case ch <- SlotData{Slot: ei.Slot, BlockNumber: ei.BlockNumber, ValidatorPubkey: vpub}:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	logger.Info("RecentMissing processed", "blocks", len(blocks))
+	return nil
+}
+
+// RecentBids backfills bid data for ALL recent slots (not just opted-in blocks)
+func recentBids(ctx context.Context, db *database.DB, httpc *retryablehttp.Client, relays []relay.Row, slots []int64) error {
+	logger := slog.With("component", "backfill")
+
+	for _, slot := range slots {
+		if ctx.Err() != nil {
+			break
+		}
+
+		for _, rr := range relays {
+			if ctx.Err() != nil {
+				break
+			}
+
+			fetchCtx, fcancel := context.WithTimeout(ctx, 5*time.Second)
+			bids, err := relay.FetchBuilderBlocksReceived(fetchCtx, httpc, rr.URL, slot)
+			fcancel()
+			if err != nil {
+				logger.Error("RecentBids fetch failed", "slot", slot, "relay_id", rr.ID, "relay_url", rr.URL, "error", err)
+				continue
+			}
+
+			rows := make([]database.BidRow, 0, len(bids))
+			for _, b := range bids {
+				if row, ok := relay.BuildBidInsert(slot, rr.ID, b); ok {
+					rows = append(rows, row)
+				}
+			}
+
+			if len(rows) > 0 {
+				insCtx, icancel := context.WithTimeout(ctx, 5*time.Second)
+				if err := db.InsertBidsBatch(insCtx, rows); err != nil {
+					icancel()
+					return fmt.Errorf("bids insert failed slot=%d relay_id=%d: %w", slot, rr.ID, err)
+				}
+				icancel()
+			}
+		}
+
+	}
+
+	logger.Info("RecentBids processed", "slots", len(slots))
+	return nil
+}
+
+// RunAll executes all backfill operations ensuring complete coverage
+func RunAll(ctx context.Context, db *database.DB, httpc *retryablehttp.Client, cfg *config.Config, relays []relay.Row) error {
+	logger := slog.With("component", "backfill")
+	logger.Info("Starting comprehensive backfill for ALL blocks (not just opted-in)")
+
+	// Channel to pass slot data from stage 1 to stages 2 & 3
+	slotChan := make(chan SlotData, cfg.BackfillBatch)
+	errCh := make(chan error, 1)
+
+	// Run recentMissing and collect slot data
+	go func() {
+		if err := recentMissing(ctx, db, httpc, cfg, cfg.BackfillLookback, cfg.BackfillBatch, slotChan); err != nil {
+			errCh <- err
+		}
+		close(slotChan)
+	}()
+
+	// Collect slots and validator data from channel
+	var slotsForBids []int64
+	var validatorsToCheck []SlotData
+
+	for data := range slotChan {
```

Comment:
Better way is to have the slotChan be a channel of some type
```
type result struct {
  data SlotData
  err Error
}
```

This way you need only 1 for loop and you can check
```
if data.err != nil {
  // handle the error
}
```